File: frontend/types/scan.ts
export interface ScanRequest {
  url: string;
  scan_source?: string;
  scan_metadata?: {
    qr_image_name?: string;
    decoded_url?: string;
    domain_intelligence?: {
      created_date: string;
      domain_age_days: number;
      registrar: string;
    };
    threat_feeds?: {
      openphish_match: boolean;
      phishtank_match: boolean;
      urlhaus_match: boolean;
      matched_sources: string[];
      confidence: string;
    };
    [key: string]: string | number | boolean | string[] | Record<string, unknown> | null | undefined;
  };
}

export interface TechnicalDetails {
  https: boolean;
  domain: string;
  subdomain_count: number;
  url_length: number;
  contains_ip: boolean;
  suspicious_keywords_found: string[];
  suspicious_tld: boolean;
  redirect_pattern_detected: boolean;
  path_depth?: number;
  query_parameter_count?: number;
  entropy_score?: number;
  scoring_breakdown?: Array<{ rule: string; points: number }>;
  rule_based_result?: {
    score: number;
    status: string;
    reasons: string[];
  };
  ml_result?: {
    score: number;
    status: string;
    reasons: string[];
    confidence?: number;
  };
  shared_indicators?: string[];
  unique_findings?: {
    rule_based: string[];
    ml: string[];
  };
  score_difference?: number;
  // Phase 9 fields
  intelligence_flags?: string[];
  is_whitelisted?: boolean;
  whitelist_reason?: string;
  is_blacklisted?: boolean;
  blacklist_source?: string;
  brand_spoof_detected?: boolean;
  suspected_brand?: string;
  spoof_explanation?: string;
  spoof_type?: string;
  feature_importances?: Array<{
    feature: string;
    label: string;
    contribution_pct: number;
    is_active: boolean;
  }>;
  ml_interpretation?: string;
  // Phase 10 fields
  threat_category?: string;
  secondary_threat_tags?: string[];
  severity_tier?: string;
  consensus_level?: string;
  educational_insight?: string;
  scan_journey?: Array<{
    stage: string;
    status: 'passed' | 'triggered' | 'warning' | 'critical' | 'informational';
    message: string;
  }>;
}


export interface ScanResult {
  scan_id: string;
  scan_type: string;
  scanned_url: string;
  status: 'SAFE' | 'SUSPICIOUS' | 'DANGEROUS';
  score: number;
  reasons: string[];
  technical_details: TechnicalDetails;
  recommendation: string;
  timestamp: string;
  confidence?: number;
  scan_source?: string;
  scan_metadata?: any;
}

// Map the DB schema as well
export interface DBScan {
  id: string;
  user_id: string;
  url: string;
  scan_type: string;
  status: string;
  score: number;
  reasons: string[];
  technical_details: TechnicalDetails | Record<string, unknown>;
  recommendation: string;
  created_at: string;
  scan_source?: string;
  scan_metadata?: any;
}

File: backend/app/schemas/scan.py
from pydantic import BaseModel, Field, field_validator
from typing import List, Dict, Any, Optional
import re

class ScanRequest(BaseModel):
    url: str = Field(
        ...,
        min_length=1,
        max_length=2048,
        description="The URL to scan. Must be between 1 and 2048 characters."
    )
    scan_source: str = Field("manual", description="Source of the scan (manual or qr).")
    scan_metadata: Dict[str, Any] = Field(default_factory=dict, description="Metadata for the scan.")

    @field_validator("url")
    @classmethod
    def reject_dangerous_schemes(cls, v: str) -> str:
        """Reject clearly dangerous or non-HTTP(S) schemes before any processing."""
        stripped = v.strip().lower()
        dangerous_prefixes = ("javascript:", "data:", "vbscript:", "file:", "about:", "blob:")
        for prefix in dangerous_prefixes:
            if stripped.startswith(prefix):
                raise ValueError(f"Dangerous URL scheme detected and rejected.")
        return v.strip()

class TechnicalDetails(BaseModel):
    https: bool
    domain: str
    subdomain_count: int
    url_length: int
    contains_ip: bool
    suspicious_keywords_found: List[str]
    suspicious_tld: bool
    redirect_pattern_detected: bool
    # ML-specific features (optional to keep backwards compatibility)
    path_depth: Optional[int] = None
    query_parameter_count: Optional[int] = None
    entropy_score: Optional[float] = None
    # Explainable scoring
    scoring_breakdown: Optional[List[Dict[str, Any]]] = None
    # Comparison-specific fields
    rule_based_result: Optional[Dict[str, Any]] = None
    ml_result: Optional[Dict[str, Any]] = None
    shared_indicators: Optional[List[str]] = None
    unique_findings: Optional[Dict[str, List[str]]] = None
    score_difference: Optional[int] = None
    model_outputs: Optional[Dict[str, Any]] = None

    # Layered Threat Intelligence (Phase 9)
    intelligence_flags: Optional[List[str]] = None
    is_whitelisted: Optional[bool] = None
    whitelist_reason: Optional[str] = None
    is_blacklisted: Optional[bool] = None
    blacklist_source: Optional[str] = None
    brand_spoof_detected: Optional[bool] = None
    suspected_brand: Optional[str] = None
    spoof_explanation: Optional[str] = None
    spoof_type: Optional[str] = None
    feature_importances: Optional[List[Dict[str, Any]]] = None
    ml_interpretation: Optional[str] = None

    # Analyst Threat Intelligence Metrics (Phase 10)
    threat_category: Optional[str] = None
    secondary_threat_tags: Optional[List[str]] = None
    severity_tier: Optional[str] = None
    consensus_level: Optional[str] = None
    educational_insight: Optional[str] = None
    scan_journey: Optional[List[Dict[str, Any]]] = None

    # Threat Intelligence Feeds (Phase 11)
    threat_feeds: Optional[Dict[str, Any]] = None



class ScanResponse(BaseModel):
    scan_id: str
    scan_type: str
    scanned_url: str
    status: str
    score: int
    reasons: List[str]
    technical_details: TechnicalDetails
    recommendation: str
    timestamp: str
    # ML-specific confidence (optional)
    confidence: Optional[str] = None
    
    # Metadata
    scan_source: str = "manual"
    scan_metadata: Dict[str, Any] = Field(default_factory=dict)

